File: src/main/java/org/dukcode/ps/codetree/trail02/chapter05/lesson03/IntroSnailNumberSquare.java
package org.dukcode.ps.codetree.trail02.chapter05.lesson03;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-snail-number-square/description">LINK</a>
 */
public class IntroSnailNumberSquare {

  // R D L U
  private static final int[] DY = {0, 1, 0, -1};
  private static final int[] DX = {1, 0, -1, 0};

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    st = new StringTokenizer(br.readLine());
    int h = Integer.parseInt(st.nextToken());
    int w = Integer.parseInt(st.nextToken());

    int[][] board = new int[h][w];

    int r = 0;
    int c = 0;
    int dir = 0;

    for (int num = 1; num <= w * h; ++num) {
      board[r][c] = num;

      int ny = r + DY[dir];
      int nx = c + DX[dir];

      if ((ny < 0 || ny >= h || nx < 0 || nx >= w) || board[ny][nx] != 0) {
        dir = (dir + 1) % 4;
      }

      r += DY[dir];
      c += DX[dir];
    }

    for (int y = 0; y < h; y++) {
      for (int x = 0; x < w; x++) {
        bw.write(String.valueOf(board[y][x]));
        bw.write(' ');
      }
      bw.newLine();
    }
    bw.newLine();

    br.close();
    bw.close();

  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter05/lesson03/IntroSnailNumberSquare.kt
package org.dukcode.ps.codetree.trail02.chapter05.lesson03.introSnailNumberSquare

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

// R, D, L, U
private val dy = intArrayOf(0, 1, 0, -1)
private val dx = intArrayOf(1, 0, -1, 0)

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-snail-number-square/description">LINK</a>
 */
fun main() {
    val (h, w) = parseInts()

    val board = Array(h) { IntArray(w) }

    fun canFillNext(
        y: Int,
        x: Int,
        dir: Int,
    ): Boolean {
        val ny = y + dy[dir]
        val nx = x + dx[dir]

        return ny in 0 until h && nx in 0 until w && board[ny][nx] == 0
    }

    var y = 0
    var x = 0
    var dir = 0
    for (num in 1..h * w) {
        board[y][x] = num

        if (!canFillNext(y, x, dir)) {
            dir = (dir + 1) % 4
        }

        y += dy[dir]
        x += dx[dir]
    }

    bw.write(board.joinToString("\n") { it.joinToString(" ") })

    bw.flush()
}

private fun parseInts() =
    StringTokenizer(br.readLine()).run {
        IntArray(countTokens()) {
            nextToken().toInt()
        }
    }
